samples: panic on bad subscription urls with gcmsg

$ gcmsg sub rabbitpubsub://guest:guest@localhost:5672/topics/test-topic
panic: runtime error: index out of range

goroutine 1 [running]:
main.parseSubscriptionURL(0x7ffeefbff0ca, 0x3b, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
	/Users/issactrotts/gocloud.dev/samples/gcmsg/url.go:88 +0x5a9
main.(*subCmd).sub(0xc0000b1368, 0x15f92e0, 0xc0000b0008, 0x7ffeefbff0ca, 0x3b, 0x15f3980, 0xc0000c0008, 0x0, 0x0)
	/Users/issactrotts/gocloud.dev/samples/gcmsg/main.go:113 +0x7a
main.(*subCmd).Execute(0xc0000b1368, 0x15f92e0, 0xc0000b0008, 0xc0000c2a20, 0x0, 0x0, 0x0, 0x1411876)
	/Users/issactrotts/gocloud.dev/samples/gcmsg/main.go:105 +0x83
github.com/google/subcommands.(*Commander).Execute(0xc0001ba770, 0x15f92e0, 0xc0000b0008, 0x0, 0x0, 0x0, 0x0)
	/Users/issactrotts/pkg/mod/github.com/google/subcommands@v0.0.0-20181012225330-46f0354f6315/subcommands.go:142 +0x2fa
github.com/google/subcommands.Execute(0x15f92e0, 0xc0000b0008, 0x0, 0x0, 0x0, 0xc000096058)
	/Users/issactrotts/pkg/mod/github.com/google/subcommands@v0.0.0-20181012225330-46f0354f6315/subcommands.go:420 +0x5f
main.main()
	/Users/issactrotts/gocloud.dev/samples/gcmsg/main.go:145 +0x135

There should be a human-friendly error message instead.

--- a/samples/gcmsg/url.go
+++ b/samples/gcmsg/url.go
@@ -82,9 +82,15 @@ func parseSubscriptionURL(s string) (URL, error) {
 	switch scheme {
 	case "gcppubsub":
 		m := gcpSubscriptionRx.FindStringSubmatch(nonScheme)
+		if len(m) != 3 {
+			return URL{}, fmt.Errorf("failed to match %s against %s", nonScheme, gcpSubscriptionRx)
+		}
 		return URL{Provider: "gcp", Project: m[1], Subscription: m[2]}, nil
 	case "rabbitpubsub":
 		m := rabbitSubscriptionRx.FindStringSubmatch(nonScheme)
+		if len(m) != 3 {
+			return URL{}, fmt.Errorf("failed to match %s against %s", nonScheme, rabbitSubscriptionRx)
+		}
 		return URL{Provider: "rabbit", ServerURL: "amqp://" + m[1], Subscription: m[2]}, nil
 	case "":
 		return URL{}, fmt.Errorf(`scheme missing from URL: "%s"`, s)

--- a/samples/gcmsg/url_test.go
+++ b/samples/gcmsg/url_test.go
@@ -39,7 +39,7 @@ func TestParseTopicURL(t *testing.T) {
 		})
 	}
 
-	badInputs := []string{"", "a", "http://www.foo.com"}
+	badInputs := []string{"", "a", "http://www.foo.com", "rabbitpubsub://guest:guest@localhost:5672/subscriptions/test-subscription-1"}
 	for _, bi := range badInputs {
 		t.Run(fmt.Sprintf("bad input: %q", bi), func(t *testing.T) {
 			_, err := parseTopicURL(bi)
@@ -70,7 +70,7 @@ func TestParseSubscriptionURL(t *testing.T) {
 		})
 	}
 
-	badInputs := []string{"", "a", "http://www.foo.com"}
+	badInputs := []string{"", "a", "http://www.foo.com", "rabbitpubsub://guest:guest@localhost:5672/topics/test-topic"}
 	for _, bi := range badInputs {
 		t.Run(fmt.Sprintf("bad input: %q", bi), func(t *testing.T) {
 			_, err := parseSubscriptionURL(bi)

